File: argon-plugins-js/shared/src/main/scala/dev/argon/plugins/js/emit/TubeEmitter.scala
package dev.argon.plugins.js.emit

import dev.argon.compiler.*
import dev.argon.compiler.module.{ModuleName, ModulePath}
import dev.argon.compiler.definitions.HasDeclaration
import dev.argon.compiler.signature.ImportSpecifier
import dev.argon.compiler.tube.ArTubeC
import dev.argon.compiler.vtable.VTableBuilder
import dev.argon.io.*
import dev.argon.plugin.*
import dev.argon.plugins.js.*
import dev.argon.plugins.js.{JSOutput, JSProgramResource, estree}
import dev.argon.util.*
import zio.*
import zio.stm.*
import zio.stream.*

private[js] trait TubeEmitter extends EmitTubeCommon {
  val tube: ArTube with HasDeclaration[true]

  def emitTube: Comp[JSOutput[context.Env, context.Error]] =
    ZIO.succeed(JSOutput(
      `package` = toDirectoryStructure(emitModule),
    ))

  private def emitModule(path: ModulePath): JSProgramResource[context.Env, context.Error] =
    new JSProgramResource.Impl[context.Env, context.Error] with Resource.WithoutFileName:
      override def asModule: Comp[estree.Program] =
        for
          importMap <- TMap.empty[ImportSpecifier, String].commit
          additionalImportMap <- TMap.empty[ModuleName, TSet[String]].commit
          arModule <- tube.module(path)
          program <-
            new ModuleEmitter {
              override val context: TubeEmitter.this.context.type = TubeEmitter.this.context
              override val options: JSOptions[context.Env, context.Error] = TubeEmitter.this.options
              override val adapter: PluginContextAdapter.Aux[context.type, JSPlugin.type] = TubeEmitter.this.adapter
              override val tube: ArTube = TubeEmitter.this.tube
              override val imports: TMap[ImportSpecifier, String] = importMap
              override val additionalImports: TMap[ModuleName, TSet[String]] = additionalImportMap
              override val module: ArModule with HasDeclaration[true] = arModule
              override protected val vtableBuilder: VTableBuilder[context.type] = TubeEmitter.this.vtableBuilder
            }.program
        yield program
    end new


  private def toDirectoryStructure[R, E](f: ModulePath => JSProgramResource[R, E]): DirectoryResource[R, E, JSProgramResource] =
    def directory(files: UStream[ModuleFile]): DirectoryResource[R, E, JSProgramResource] =
      new DirectoryResource[R, E, JSProgramResource] with Resource.WithoutFileName {
        override def contents: ZStream[R, E, DirectoryEntry[R, E, JSProgramResource]] =
          val tmp: ZStream[R, E, DirectoryEntry[R, E, JSProgramResource]] =
            files.groupBy {
              case ModuleFile(head +: tail, file, path) => ZIO.succeed((Some(head), ModuleFile(tail, file, path)))
              case file => ZIO.succeed((None, file))
            } {
              case (Some(name), files) =>
                ZStream(DirectoryEntry.Subdirectory(name, directory(files)))

              case (None, files) =>
                files.map { case ModuleFile(_, file, path) => DirectoryEntry.File(file + ".js", f(path)) }
            }

          tmp
      }

    directory(ZStream.fromIterable(tube.modulePaths.iterator.map(getModuleFileName(tube)).toSeq))
  end toDirectoryStructure

}

File: plugin/js/src/emit/tube_emitter.ts
import type { DirectoryEntry, DirectoryResource } from "@argon-lang/plugin-api/resource";
import type { EmitTubeInfo, Metadata, ModuleName, ModulePath, TubeName } from "@argon-lang/plugin-api/tube";
import type { Program } from "estree";
import type { ExternImpl } from "../extern.js";
import type { JSOptions, JSOutput } from "../options.js";
import { JSProgramResource } from "../resource.js";
import { EmitTubeCommon } from "./emit_tube_common.js";
import { ModuleEmitter } from "./module_emitter.js";
import { getModuleFileName, ModuleFile } from "./path.js";



class TubeDirectoryStructure implements DirectoryResource<JSProgramResource> {
    constructor(files: () => Promise<readonly ModuleFile[]>, emitModule: (modulePath: ModulePath) => JSProgramResource) {
        this.#files = files;
        this.#emitModule = emitModule;
    }

    readonly #files: () => Promise<readonly ModuleFile[]>;
    readonly #emitModule: (modulePath: ModulePath) => JSProgramResource;

    resourceType: "directory" = "directory";

    get fileName(): string | null {
        return null;
    }

    async *contents(): AsyncIterable<DirectoryEntry<JSProgramResource>> {
        const subdirs = new Map<string, ModuleFile[]>();
        for(const file of await this.#files()) {
            const [ subdir, ...restPath ] = file.dir;
            if(subdir !== undefined) {
                let subdirFiles = subdirs.get(subdir);
                if(subdirFiles === undefined) {
                    subdirFiles = [];
                    subdirs.set(subdir, subdirFiles)
                }

                subdirFiles.push({
                    dir: restPath,
                    file: file.file,
                    path: file.path,
                });
            }
            else {
                yield {
                    entryType: "file",
                    name: file.file,
                    resource: this.#emitModule(file.path),
                };
            }
        }

        for(const [dirName, files] of subdirs) {
            yield {
                entryType: "subdirectory",
                name: dirName,
                resource: new TubeDirectoryStructure(async () => files, this.#emitModule),
            };
        }
    }

}

class TubeModuleResource extends JSProgramResource {
    constructor(emitter: TubeEmitter, path: ModulePath) {
        super();
        this.#emitter = emitter;
        this.#path = path;
    }

    readonly #emitter: TubeEmitter;
    readonly #path: ModulePath;

    get fileName(): string | null {
        return null;
    }

    async asModule(): Promise<Program> {
        const moduleDef = await this.#emitter.tube.getModule(this.#path);
        const metadata = this.#emitter.metadata;
        const moduleName: ModuleName = {
            tube: metadata.name,
            path: this.#path,
        };
        const moduleEmitter = new ModuleEmitter(this.#emitter.options, this.#emitter.emitTubeInfo, metadata, moduleName, moduleDef);

        return await moduleEmitter.program();
    }

}

export class TubeEmitter extends EmitTubeCommon {
    constructor(options: JSOptions, emitInfo: EmitTubeInfo<ExternImpl, ExternImpl, ExternImpl>, metadata: Metadata) {
        super(options, emitInfo, metadata);
    }

    async emitTube(): Promise<JSOutput> {
        return {
            package: this.#toDirectoryStructure(path => this.#emitModule(path)),
        };
    }

    #emitModule(path: ModulePath): JSProgramResource {
        return new TubeModuleResource(this, path);
    }

    #toDirectoryStructure(f: (modulePath: ModulePath) => JSProgramResource): DirectoryResource<JSProgramResource> {
        return new TubeDirectoryStructure(async () => (await this.tube.allModulePaths()).map(getModuleFileName), f);
    }

}
